src/base-implementation.ts

```diff
@@ -0,0 +1,103 @@
+import { Wallet } from '@harmony-js/account'
+import { Contract } from '@harmony-js/contract'
+import { Harmony } from '@harmony-js/core'
+import { BigNumber, logger, constants } from 'ethers'
+import { Logger } from 'ethers/lib/utils'
+import { Base } from './base-token'
+
+export class BaseError extends Error {
+  public readonly type: string
+  public readonly code: number
+  public readonly data: string
+
+  constructor(message: string, type: string, code: number, data: string) {
+    super(message)
+    this.name = BaseError.name
+    this.type = type
+    this.code = code
+    this.data = data
+
+    Error.captureStackTrace(this, this.constructor)
+  }
+}
+
+export class BaseToken extends Base {
```

Comment:
let's make this class abstract and not extend anything

Reply:
Okay!

Reply:
Okay!